---
title: "RNA-Seq Workshop in R"
output: html_document
---

```{r}
lib.loc <- "/homedirs18/sghms/bms/shares/bdib/BDiB_2018_19/R/packages_r_3_5_0"
stopifnot(file.exists(lib.loc))
.libPaths()
.libPaths(lib.loc)
.libPaths()
library(magrittr, lib.loc=lib.loc)
library(dplyr, lib.loc=lib.loc)
```

### STEP 3: Load the following R packages into R by copying and pasting all the following at the terminal

These packages are required for us to perform our analysis steps in DESeq2, and some additional packages to manipulate our data and draw graphs.

```{r}
library("Rsamtools")
library("GenomicFeatures")
library("GenomicAlignments")
library("backports")
library("DESeq2")
library("ggplot2")
library("org.Hs.eg.db")
library("gplots")
```


library("pheatmap")
library("RColorBrewer")

### STEP 4: Type the following into the console to identify the working current R working directory:

```{r}
getwd()
```


### STEP 7: Check the downloaded and unpacked file "CountsTable_PPP2R3B.txt" is present in the working directory 

Type the following into the console:

```{r}
list.files()
```

Q. Can you see that the counts table is one of the listed files?

## PART 2 : Preparing The Count Data For DESeq2 analysis

### STEP 8: Loading the count data into R 

Here we are loading the RNAseq count data into R as a matrix object. Type the following into the R console:

```{r}
CountsTable_PPP2R3B <- as.matrix(read.table("CountsTable_PPP2R3B.txt", header = TRUE))
```

### STEP 9: View the top of the "CountsTable_PPP2R3B.txt" matrix using the head command in R:

```{r}
head(CountsTable_PPP2R3B)
```

Q. Take a moment and look at how the count data is arranged; the first column is the genomic feature (in our case a gene exon), the subsequent 3 columns are the RNA read counts of each of our untreated cell line samples. The next three columns are the RNA read counts of each of our treated cell lines.


### STEP 10: Creating our DESeq experiment


```{r}
DESeq2Experiment <- as.matrix(CountsTable_PPP2R3B)
```

Here we are assigning untreated or treated status to each of cell lines columns of count data:
"condition". The following lines of R code will build our DESeq2Experiment object with our cell line conditions:

```{r}
condition <- factor(c(rep("untreated", 3), rep("treated", 3)))
```

```{r}
coldata <- data.frame(row.names=colnames(DESeq2Experiment), condition)
```

Next we will set our DESeq2 experiment to our the "condition" of or cells as defined above

```{r}
DESeq2Experiment <- DESeqDataSetFromMatrix(countData=DESeq2Experiment, colData=coldata, design=~condition)
```
This means that in our DESeq2 analysis we will be comparing the untreated condition of our cells to the treated condition of our cells.

We also need to check that our base reference is the "untreated" sample: 

```{r}
DESeq2Experiment$condition
```

If not we can re-level the order with the following function- this is because we want out untreated cells to be at the base level when we perform the differential gene expression analysis.

```{r}
DESeq2Experiment$condition <- relevel(DESeq2Experiment$condition, "untreated")
```

and now just checking again that the level ordering has been changed:

```{r}
DESeq2Experiment$condition
```

Before we conduct our differential gene expression analysis, we will perform some quality control procedures to assess the quality of our data. 

## Part 3 : Quality Control

### STEP 11: Scatter plots

The function below rlog returns an object containg the rlog-transformed values  

```{r}
rld <- rlog(DESeq2Experiment, blind = FALSE)
```

To compare the rlog transformed values of sample 1 (SKMEL2_Untreated_1) to sample 2 (SKMEL2_Untreated_1) use the following R code

```{r}
plot(assay(rld) [,1:2], pch=16, cex=0.3)
```

Q. Modify the R code above to produce another graph comparing sample 4 to sample 6

Q. Why are we interested in comparing our experimental samples?

### STEP 12: Principle Component Analysis (PCA)

With a PCA analysis we can visualise how close (or far) our experimental samples cluster from one another. In normal experimental conditions we would like our sample replicates to not be too dissimilar from each other.

The following function in R will generate a PCA plot:

```{r}
plotPCA(rld, intgroup="condition")
```
Q. Take a moment to look at the PCA plot. What is this analysis showing us?

Heatmap of sample-to-sample distances
```{r}
sampleDists <- dist( t( assay(rld) ) )
sampleDists
```


sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- condition
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
	  clustering_distance_rows=sampleDists,
	  clustering_distance_cols=sampleDists,
	  col=colors)



## Part 4 : Differential Gene Expression Analysis in DESeq2


### STEP 13: Run the differential gene expression analysis in DESeq2:

Here we are calling the major DESeq2 function which is to perform the differential gene expression analysis:

```{r}
DESeq2Experiment <- DESeq(DESeq2Experiment)
```

Now with the results function we will build a table of the differential gene expression analysis results

```{r}
results <- results(DESeq2Experiment)
```

View the results table with the head function in R

```{r}
head(results)
```

Q. Take a moment to look at the results table. What information has DESeq2 reported? What additional information would we like to have alongside our results? 

## Part 5 : Reporting and Data Visualisation 

### STEP 14: Reporting the differential gene expression results

Annotate with gene IDs (gene symbol and EntrezID) using the database package #org.Hs.eg.db. We want to know what gene is which ! So we can populate our results table with gene level information.

First annotate the results with the gene symbol

```{r}
results$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(results),
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")
```

Check that the gene symbol has been added to the results table with the head function

```{r}
head(results)
```

Now annotate the results with the entrez ID number

```{r}
results$entrez <- mapIds(org.Hs.eg.db,
                   keys=row.names(results),
                   column="ENTREZID",
                   keytype="ENSEMBL",
                   multiVals="first")
```

Check that the entrez ID has been added to the results table with the head function

```{r}
head(results)
```

### STEP 15: Writing the results to an excel file

We can order our results by the adjusted pvalues so that the most significant results are at the top:

```{r}
resOrdered <- results[order(results$padj), ]
```

Q. What are adjusted pvalues?

We can then view this new ordering with the head function

```{r}
head(resOrdered)
```

It's useful to be able to export our results to an excel file for ease of viewing the data.

First we will convert our results to a data frame

```{R}
resOrderedDF <- as.data.frame(resOrdered)
```

Here we name our excel file (name the file anything that you would like!)

```{r}
outputCSV <- ('Differential_expression_results.csv')
```

Here we call the R function write.csv to convert our results data frame into an excel file 

```{r}
write.csv(resOrderedDF, file = outputCSV)
```

Q. Did it create the excel file? go to the folder and open the excel file. What are the top 2 significant results? is PPP2R3B there as expected? are there any other interesting findings?

### STEP 16: MA - Plot

An MA-plot provides a useful overview for an experiment with a two-group comparison. Significant results are in red, non-significant results are in black

An MA-plot can be generated with the following function in R

```{r}
plotMA(results, 
	ylim=c(-5,5))
```
	
Q. Why is there a large spread of non-significant results on the left hand side of the MA-plot?	

### STEP 17: Volcano - Plot

A Volcano plot is a nice way of presenting differential gene expression analysis from RNAseq data.

In statistics, a volcano plot is a type of scatter-plot that is used to quickly identify changes in large data sets composed of replicate data. It plots significance versus fold-change on the y and x axes, respectively. 

```{r}
results$sig <- -log10(results$padj)
```

```{r}
sum(is.infinite(results$sig))
```

```{r}
results[is.infinite(results$sig),"sig"] <- 350
```

Select genes with a defined p-value (DESeq2 assigns NA to some genes and we want to exclude any of these)

```{r}
genes.to.plot <- !is.na(results$pvalue)
```

```{r}
range(results[genes.to.plot,
		"log2FoldChange"])
```

Volcano plot of adjusted p-values

```{r}
alpha <- 0.05
```

```{r}
cols <- densCols(results$log2FoldChange, results$sig)
cols[results$pvalue ==0] <- "purple"
results$pch <- 19
results$pch[results$pvalue ==0] <- 6
plot(results$log2FoldChange, 
     results$sig, 
     col=cols, panel.first=grid(),
     main="Volcano plot", 
     xlab="Effect size: log2(fold-change)",
     ylab="-log10(adjusted p-value)",
     pch=results$pch, cex=0.4)
abline(v=0)
abline(v=c(-1,1), col="brown")
abline(h=-log10(alpha), col="brown")
```

Q. What is the volcano plot showing us with respect to the differential gene expression results?

#################################

### STEP 17: Volcano - Plot 2

lets remove our two most significant genes to scale our graph better, here we remove the top 4 most significant SNPs
```{r}
resOrderedDF_mod <- resOrderedDF[-1:-4,]
```



```{r}
resOrderedDF_mod$sig <- -log10(resOrderedDF_mod$padj)
```

```{r}
sum(is.infinite(resOrderedDF_mod$sig))
```

```{r}
resOrderedDF_mod[is.infinite(resOrderedDF_mod$sig),"sig"] <- 350
```

Select genes with a defined p-value (DESeq2 assigns NA to some genes and we want to exclude any of these)

```{r}
genes.to.plot <- !is.na(resOrderedDF_mod$pvalue)
```

```{r}
range(resOrderedDF_mod[genes.to.plot,
		"log2FoldChange"])
```

Volcano plot of adjusted p-values

```{r}
alpha <- 0.05
```

```{r}
cols <- densCols(resOrderedDF_mod$log2FoldChange, resOrderedDF_mod$sig)
cols[resOrderedDF_mod$pvalue ==0] <- "purple"
resOrderedDF_mod$pch <- 19
resOrderedDF_mod$pch[resOrderedDF_mod$pvalue ==0] <- 6
plot(resOrderedDF_mod$log2FoldChange, 
     resOrderedDF_mod$sig, 
     col=cols, panel.first=grid(),
     main="Volcano plot", 
     xlab="Effect size: log2(fold-change)",
     ylab="-log10(adjusted p-value)",
     pch=results$pch, cex=0.4)
abline(v=0)
abline(v=c(-1,1), col="brown")
abline(h=-log10(alpha), col="brown")
```

### How many differentially expressed genes ?

```{r}
sig <- resOrdered[!is.na(resOrdered$padj) & resOrdered$padj<0.1 & abs(resOrdered$log2FoldChange)>=0.5,]
```

```{r}
head(sig)
```

### Heat Map

```{r}
selected <- rownames(sig);selected
```


colours of heatmap

```{r}
hmcol <- heat.colors(5, alpha = 1)
```

```{r}
graphics.off()
```

```{r}
par("mar")
```

```{r}
par(mar=c(1,1,1,1))
```


 heatmap.2( log2(counts(DESeq2Experiment,normalized=TRUE) [rownames(DESeq2Experiment) %in% selected,]), col = hmcol, scale = "row", Rowv = TRUE, Colv = FALSE, dendrogram = "row", trace = "none", margin=c(4,6), cexRow = 0.4, cexCol = 1, keysize = 0.8)




#Workshop complete !